### DISCUSSION TITLE
How to address structure offset intentions correctly

### DISCUSSION THREAD

**Discussion by hellow554** [RPGHacker/asar#321] [Q&A] [answered]
I try to "decompile" a popular SNES game and often the code looks something like this:

```asm
struct Player_t $8000
  .x: skip 1
  .y: skip 1
endstruct

LDX.W #Player_t ; base address of an struct array
LDA.W $01, X ; use the second byte of that struct
```

Of course these don't stand next to each other and often it is assumed that `X` points to an struct in that array.

I can think of two possibilities. What other ideas do you have in mind?

 1. This one would make the lowest effort to change something (just "invent" an `offsetof` macro or function): `LDA.W offsetof(player.Y), X`
 2. This one would require more modification to the base, but makes then intention much clearer IMHO: `LDA.W player[X].y`, but I'm not sure how other would find this, because it lacks the `, X` in the end which has moved into the array index. Second `X` is now not an array index, but an offset from `$0`.

What do you have mind to make the intention more clear that the original example?

**Comment by RPGHacker** [comment, marked as answer]
An `offsetof()` does sound like a reasonable addition to Asar (though I'm not sure how easy it is to add - don't remember how exactly struct labels are stored in Asar).

For the time being, I'd propose a workaround similar to this:

```
PlayerData = $8000
struct Player_t
  .x: skip 1
  .y: skip 1
endstruct

LDX.W #PlayerData ; base address of an struct array
LDA.W Player_t.y, X ; use the second byte of that struct
```

**Reply by hellow554** [reply]
I didn't know that you can omit the address on a struct. [The documentation](https://rpghacker.github.io/asar/asar_2_beta/structs.html) says otherwise ;)
I'll try it out and get back to you

**Reply by hellow554** [reply]
```
common/structs.asm:31: error: (Ebank_border_crossed): A bank border was crossed, SNES address $000001.
    in block: [.pos_x: skip 2]
```

Because the `structs.asm` file is included becfore the first `ORG` or `BASE` command, the error pops up.

I inserted a dummy `BASE $8000` and it works now, but it feels hacky....

**Reply by RPGHacker** [reply]
I thought it was a thing, but maybe I misremembered. If it doesn't work, I guess you could also do

```
struct Player_t $000000
  .x: skip 1
  .y: skip 1
endstruct
```

It's what I use in VWF Dialogues, and it doesn't seem to throw bank border errors. Example can be seen here:
https://github.com/RPGHacker/SMW-Workspace/blob/master/patches/shared/shared.asm
in the `define_enum()` macro.
